### ISSUE TITLE
messages getting lost when autoCommitOffset is set to true and autoCommitOnError to false

### ISSUE THREAD

**Issue by VasuBhaskar** [spring-cloud/spring-cloud-stream-binder-kafka#468]
We had the setting as below

            ```
  autoRebalanceEnabled: true
              autoCommitOffset: true
              maxPollRecords: 1
              autoCommitOnError: false
```

There was an intermittent error while consuming messages as below

```
{
  "@timestamp": "2018-10-12T03:24:29.676+00:00",
  "@version": "1",
  "message": "Exception thrown when sending a message with key='xxxx' and payload='{\"key\":\xxxxxxx\",\"source_system\":\"xxxx\",\"timestamp\":\"2018-10-11T22:24:29.000+0000\",\"d...' to topic xxxxxxxxx:",
  "logger_name": "org.springframework.kafka.support.LoggingProducerListener",
  "thread_name": "kafka-producer-network-thread | producer-2",
  "level": "ERROR",
  "level_value": 40000,
  "stack_trace": "org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.\n"
}
```

After this exception, the lag from the consumer group got reduced consistently, but no new messages were processed(available in the log).

Initially had autoCommitOnError  also as true and we thought that is what causing the issue and changed it to autoCommitOnError as false. But still the same issue.

  autoRebalanceEnabled: true
              autoCommitOffset: true
              maxPollRecords: 1
              autoCommitOnError: true

consumer code as below

```
@StreamListener
    @Output(XXXXXX)
    Flux<Message<String>> processMessage(@Input(XXXXXX.INPUT) Flux<String> events) {
        events
            .log()
            .map { it -> decorateEvent() }
            .map { it -> buildMessage(it.key, objectMapper.writeValueAsString(it)) }
            .doOnError{ log.info('Exception while sending the record  ' + it + ' - ' + it.getCause()) }
    }
```

We are trying to do a manual commit now, but not sure if that will also fix the issue. any quick response will be much appreciated!

**Comment by VasuBhaskar** [comment]
@sobychacko Any quick help will be much appreciated. Thanks a lot for looking in to this

**Comment by VasuBhaskar** [comment]
@olegz could you please have a look in to this

**Comment by olegz** [comment]
@VasuBhaskar please stop tagging individual contributors expecting a quick reply. It's not helpful not it is fair. If this is a production issue and you are a customer of Pivotal and need a priority involvement please contact our support organization. Otherwise, please be patient and we'll get to this and other issues in due time.

**Comment by sobychacko** [comment]
@VasuBhaskar Could you provide a minimal sample application and the configuration where we can reproduce this issue? That way we can troubleshoot the issue faster. Without that, it is hard to say what is happening.

**Comment by VasuBhaskar** [comment]
@sobychacko am very sorry to have tagged individual guys :-( . Just felt that urgency as it was a prod issue. This issue occurs only when there is an connection exception while sending to topic or reading from the topic . Otherwise everything else works just fine. After the exception, offsets get committed, even though nothing gets processed

**Comment by sobychacko** [comment]
@VasuBhaskar No worries. So if I understand it correctly, there are no consistent ways to reproduce this issue, but only happens while there are connection issues. Is that right? I will investigate further if there are any known issues.

**Comment by VasuBhaskar** [comment]
@sobychacko Yes. That is right. And this is the exception that occurred as well which created the problem in the first place.

```
{
  "@timestamp": "2018-10-12T03:24:29.676+00:00",
  "@version": "1",
  "message": "Exception thrown when sending a message with key='xxxx' and payload='{\"key\":\xxxxxxx\",\"source_system\":\"xxxx\",\"timestamp\":\"2018-10-11T22:24:29.000+0000\",\"d...' to topic xxxxxxxxx:",
  "logger_name": "org.springframework.kafka.support.LoggingProducerListener",
  "thread_name": "kafka-producer-network-thread | producer-2",
  "level": "ERROR",
  "level_value": 40000,
  "stack_trace": "org.apache.kafka.common.errors.NotLeaderForPartitionException: This server is not the leader for that topic-partition.\n"
}
```

do u think manual ack will resolve this problem.

**Comment by sobychacko** [comment]
Can you turn `DEBUG` logging on and get more stack trace output?

**Comment by sobychacko** [comment]
@VasuBhaskar and also, what kind of connections issues the producer is facing? Try the debug logging. It sounds to me that this problem maybe reproducible even if by simulating some connection issues on the outbound. See if you can try that.

**Comment by VasuBhaskar** [comment]
@sobychacko working on simulating some connection issues on the outbound. will keep you posted

**Comment by sobychacko** [comment]
@VasuBhaskar Is there an update to this issue? Were you able to resolve the issue?
